fix(volume_xfs): recreating a volume with quota skips project id assignment

debdutdeb opened the pull request:
Signed-off-by: Debdut Chakraborty <debdutdeb@outlook.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->
Closes #43891 

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
* Remove volume path from quota state if respective volume is being deleted


**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/76006232/183312172-4fef064f-79c7-45e7-a391-1e297cd24a2b.png)

debdutdeb committed:
[projectquota] remove path from state if volume is deleted

Signed-off-by: Debdut Chakraborty <debdut.chakraborty@rocket.chat>

vvoland reviewed (commented):
Thanks! I don't know much about the xfs quota but left some comments for things that drew my attention.

debdutdeb committed:
[projectquota] use quota state map appropriately

This keeps the projectId and volume path in memory for the daemon's lifetime.
If the same volume is created again, reuse the id. Else create a new id.

Signed-off-by: Debdut Chakraborty <debdut.chakraborty@rocket.chat>

debdutdeb committed:
[gitignore] remove .idea/ from .gitignore

Signed-off-by: Debdut Chakraborty <debdut.chakraborty@rocket.chat>

debdutdeb commented on quota/projectquota.go:
```diff
@@ -316,6 +316,12 @@ func setProjectID(targetPath string, projectID uint32) error {
 
+func (q *Control) RemoveQuota(targetPath string) {
+	q.Lock()
+	delete(q.quotas, targetPath)
```
No need to do that. 

But, here removing the path from the map was causing adding more projetIds for the same path (e.g. `docker create volume -o size=10m abc # #1001`, `docker volume rm abc`, `docker create volume -o size=10m abc # #1002`...). Wasn't ideal. 

I have updated the patch to reuse the same project id for the same volume for any continuous runtime (e.g. `docker create volume -o size=10m abc # #1001`, `docker volume rm abc`, `docker create volume -o size=10m abc # #1001`...). 

There is still the case where a project id will stay stale if someone removes a volume with quota and create another with a diff name, and so on. Only can be rectified (i.e. reuse those ids) by restarting the daemon. 

Unfortunately I'm on a mac, and to test and run this I had to use a linux server, connect remotely and what not. But should be a matter of extending the state to reflect whether is allocated or not, and toggle it at different events. I am thinking of handling that problem separately once my linux system gets repaired.

debdutdeb commented:
/ping @vvoland @thaJeztah =)

debdutdeb commented:
Friendly ping @thaJeztah @vvoland

cpuguy83 commented on quota/projectquota.go:
```diff
@@ -195,23 +195,30 @@ func (q *Control) SetQuota(targetPath string, quota Quota) error {
-		//
-		err := setProjectID(targetPath, projectID)
-		if err != nil {
+		defer func() {
```
I don't understand this change.
The only difference seems to be before we incremented the project ID before setting the quota, now it increments after and I don't see how that would change anything significantly.

debdutdeb replied:
Not exactly. Previously if we created the same named volume the second time, because the path was still in the map with a project id, it skipped actually setting the project id to the path, which then caused the quota to not actually get set. 

Previously
```go
        projectID, ok := q.quotas[targetPath]
	if !ok {
                // a new path/volume; so need a new project id
		state := getPquotaState()
		projectID = state.nextProjectID

		//
		// assign project id to new container directory
		//
		err := setProjectID(targetPath, projectID)
		if err != nil {
			state.Unlock()
			return err
		}

		state.nextProjectID++

                // save to state
		q.quotas[targetPath] = projectID
	}
        // here means the path/volume was in state
        // but disregards for the case where the volume might've been deleted at some point, 
        // invalidating/freeing the previously set project id.
```

Now, 

```go
	projectID, ok := q.quotas[targetPath]
	if !ok {
                // a new volume; get a new project id
		state := getPquotaState()
		state.Lock()
		projectID = state.nextProjectID

                 // Not safe to increment the project id just yet, in case setProjectId fails
		defer func() {
			if _, exists := q.quotas[targetPath]; exists {
				state.nextProjectID++
			}
			state.Unlock()
		}()
	}


        // Reuse the project id if found in state
        // or the new one from the if block above
	//
	// assign project id to new container directory
	//
	err := setProjectID(targetPath, projectID)
	if err != nil {
		return err
	}

	q.quotas[targetPath] = projectID
```

Please let me know if im missing something.

debdutdeb replied:
Another bit of background [here](https://github.com/moby/moby/pull/43928?notification_referrer_id=NT_kwDOBIfDWLM0MTUwNTE3Njc4Ojc2MDA2MjMy#discussion_r940613731)

PS: hasn't been repaired yet.

debdutdeb commented:
Hi @thaJeztah do you mind giving this another check? Since the issue is still open.
